### PR TITLE
DT-5921 Add car ferry functionality

### PR DESCRIPTION
**Pull request by VillePihlava**
This should not be merged before the functionality from this pr can be used https://github.com/opentripplanner/OpenTripPlanner/pull/5966,

## Proposed Changes

  - Add car ferry functionality
    - matka config
    - waltti config
  - Direct car and cars on transit combination view
  - Add overflow icon
  - Remove leg text for itineraries for ferries in the matka config
  - Fix leg-before-line width for bikes from 6px to 8px
  - Fix color for the secondary icon in the alternative itinerary bar

### CODE REVIEW

File: app/component/RouteNumberContainer.js [hunk 1]
```diff
@@ -25,8 +26,15 @@ const RouteNumberContainer = (
       isCallAgency={isCallAgency || route.type === 715}
       color={route.color ? `#${route.color}` : null}
       mode={mode !== undefined ? mode : route.mode}
-      text={getLegText(route, config, interliningWithRoute)}
+      text={
+        config.disabledLegTextModes !== undefined &&
+        config.disabledLegTextModes.includes(route.mode) &&
```
Comment (vesameskanen) [inline app/component/RouteNumberContainer.js]: Use config.disabledLegTextModes?.includes()

File: app/component/RouteNumberContainer.js [hunk 2]
```diff
@@ -41,6 +49,7 @@ RouteNumberContainer.propTypes = {
   className: PropTypes.string,
   fadeLong: PropTypes.bool,
   withBicycle: PropTypes.bool,
+  withCar: PropTypes.bool,
```
Comment (vesameskanen) [inline app/component/RouteNumberContainer.js]: No need to expose withCar. Please refactor this component so that only props which are referred are declared here. Others can flow through in ...props.

File: app/component/itinerary/CarLeg.js [hunk 3]
```diff
@@ -19,6 +21,24 @@ export default function CarLeg(props, { config, intl }) {
   const firstLegClassName = props.index === 0 ? 'first' : '';
   const modeClassName = 'car';
 
+  let circleLine = (
```
Comment (vesameskanen) [inline app/component/itinerary/CarLeg.js]: const circleLine = props.carBoardingLeg  ? 
  <ItineraryCircleLineLong ...> :
  <ItineraryCircleLineWithIcon ...>;

File: app/component/itinerary/CarLeg.js [hunk 4]
```diff
@@ -70,6 +86,26 @@ export default function CarLeg(props, { config, intl }) {
             focusAction={props.focusAction}
           />
         </div>
+        {props.carBoardingLeg?.from.stop && (
+          <div className={cx('itinerary-leg-action', 'car')}>
+            <div className="itinerary-leg-action-content">
```
Comment (vesameskanen) [inline app/component/itinerary/CarLeg.js]: Merge these into one div

File: app/component/itinerary/CarLeg.js [hunk 5]
```diff
@@ -87,6 +123,26 @@ export default function CarLeg(props, { config, intl }) {
             />
           </div>
         </div>
+        {props.carBoardingLeg?.to.stop && (
+          <div className={cx('itinerary-leg-action', 'car')}>
+            <div className="itinerary-leg-action-content">
```
Comment (vesameskanen) [inline app/component/itinerary/CarLeg.js]: Merge divs

File: app/component/itinerary/InterlineInfo.js [hunk 6]
```diff
@@ -18,11 +18,15 @@ const InterlineInfo = ({ legs, leg }) => {
       }
     });
   }
+  let icon = 'icon-icon_wait';
+  if (usingOwnCarWholeTrip) {
+    icon = 'icon-icon_wait-car';
+  }
```
Comment (vesameskanen) [inline app/component/itinerary/InterlineInfo.js]: Again, use conditional assignment:
const icon = usingOwnCarWholeTrip ? wait-car : wait ;

File: app/component/itinerary/Itinerary.js [hunk 7]
```diff
@@ -277,6 +280,9 @@ const Itinerary = (
   );
   const usingOwnBicycleWholeTrip =
     usingOwnBicycle && itinerary.legs.every(leg => !leg.to?.vehicleParking);
+  const usingOwnCar = itinerary.legs.some(leg => getLegMode(leg) === 'CAR');
+  const usingOwnCarWholeTrip =
+    usingOwnCar && itinerary.legs.every(leg => !leg.to?.vehicleParking);
```
Comment (vesameskanen) [inline app/component/itinerary/Itinerary.js]: can leg.to actually be null?

File: app/component/itinerary/ItineraryDetails.js [hunk 8]
```diff
@@ -133,21 +135,30 @@ class ItineraryDetails extends React.Component {
...
     const fares = getFaresFromLegs(itinerary.legs, config);
     const extraProps = this.getExtraProps(itinerary);
     const {biking, walking, driving, futureText, isMultiRow} = extraProps;
+    // This does not take into account if the user is using a car at the time of using transit,
+    // instead this just calculates if the car is used for the whole trip
+    // a smarter approach would be to store the current personal mode of transport (walk, bike, car)
+    // this could then be used to set the waiting icon legs that need it.
+    const usingOwnCarWholeTrip = ((walking && walking.distance <= 0) || walking === undefined) &&
+                                 ((biking && biking.distance <= 0) || biking === undefined) &&
+                                 (driving && driving.distance > 0)
```
Comment (vesameskanen) [inline app/component/itinerary/ItineraryDetails.js]: It is easy to see that walking, biking and driving are always defined. So it is necessary to test only  

walking.distance === 0 && biking.distance === 0 && driving.distance > 0

File: app/component/itinerary/ItineraryListHeader.js [hunk 9]
```diff
@@ -31,6 +32,25 @@ export default function ItineraryListHeader({
...
+              className="inline-icon"
+              color="#007AC9"
+              severityLevel={AlertSeverityLevelType.Info}
+            />
+          </div>
+          <div className="with-car-info-notification">
+            <div className="with-car-info-content">
+              <FormattedMessage
+                id="itinerary-summary.car-boarding-information"
+                defaultMessage="Invalid boarding information: translation not found"
```
Comment (vesameskanen) [inline app/component/itinerary/ItineraryListHeader.js]: Convention is not to include error messages as defaultMessage. I suggest leaving the default message empty. 

We should discuss and decide a smart policy for translations.

File: app/util/legUtils.js [hunk 10]
```diff
@@ -593,17 +593,68 @@ export function getExtendedMode(leg, config) {
...
+/**
+ * Determines whether to show a notification for a car with public transit
+ *
+ * @param {object} leg - The leg object.
+ * @param {object} config - Config data.
+ * @returns {boolean} - Returns true if a notifier should be shown
+ */
+export const showCarBoardingNote = (leg, config) => {
+  const { carBoardingModes } = config;
+  return carBoardingModes && carBoardingModes[leg.mode]?.showNotification;
```
Comment (vesameskanen) [inline app/util/legUtils.js]: use ? here as well

File: app/util/legUtils.js [hunk 11]
```diff
@@ -593,17 +593,68 @@ export function getExtendedMode(leg, config) {
...
  * @param {object} leg - The leg object.
  * @param {object} config - Config data.
  * @returns {boolean} - Returns true if a notifier should be shown
  */
 export const showBikeBoardingNote = (leg, config) => {
   const { bikeBoardingModes } = config;
-  return (
-    bikeBoardingModes && bikeBoardingModes[leg.mode]?.showNotification === true
-  );
+  return bikeBoardingModes && bikeBoardingModes[leg.mode]?.showNotification;
```
Comment (vesameskanen) [inline app/util/legUtils.js]: return bikeBoardingModes?.[]

File: app/util/legUtils.js [hunk 12]
```diff
@@ -593,17 +593,68 @@ export function getExtendedMode(leg, config) {
...
+      duration: 0,
+      start: leg.start,
+      end: leg.start,
+      distance: -1,
+      rentedBike: leg.rentedBike,
+      to,
+      from,
+      mode: boardingMode,
+    };
+    return boardingLeg;
```
Comment (vesameskanen) [inline app/util/legUtils.js]: How about return {  } ? There is no need to assing to a constant first.

File: app/component/itinerary/ItineraryPageUtils.js [hunk 13]
```diff
@@ -468,6 +468,36 @@ export function mergeBikeTransitPlans(bikeParkPlan, bikeTransitPlan) {
   };
 }
 
+/**
+ * Merge the direct car plan with the car transit plan.
+ */
+export function mergeCarDirectAndTransitPlans(carPlan, carTransitPlan) {
+  let carPlanEdges = carPlan?.edges;
```
Comment (vesameskanen) [inline app/component/itinerary/ItineraryPageUtils.js]: Simpler way: carPlan?.edges || []

File: app/component/itinerary/ItineraryPageUtils.js [hunk 14]
```diff
@@ -468,6 +468,36 @@ export function mergeBikeTransitPlans(bikeParkPlan, bikeTransitPlan) {
...
+  let carPublicEdges = carTransitPlan?.edges;
+  if (carPlanEdges === undefined) {
+    carPlanEdges = [];
+  }
+  if (carPublicEdges === undefined) {
+    carPublicEdges = [];
+  }
+
+  // If there is a direct car plan.
+  if (carPlanEdges.length === 1) {
```
Comment (vesameskanen) [inline app/component/itinerary/ItineraryPageUtils.js]: The same object can be returned in both cases below, just use carPlanEdges.length as count value.

File: app/component/itinerary/ItineraryList.js [hunk 15]
```diff
@@ -111,6 +116,34 @@ function ItineraryList(
...
+        ) || 'ferry';
+      const legs = planEdges
+        .slice(carDirectItineraryCount)
+        .flatMap(edge => edge.node.legs);
+      const showCarBoardingInfo = legs.some(leg =>
+        showCarBoardingNote(leg, config),
+      );
+
+      summaries.splice(
+        carDirectItineraryCount || 0,
```
Comment (vesameskanen) [inline app/component/itinerary/ItineraryList.js]: no need for || 0

File: app/component/itinerary/ItineraryListHeader.js [hunk 16]
```diff
@@ -31,6 +32,25 @@ export default function ItineraryListHeader({
...
+      {showCarBoardingInfo && (
+        <div className="with-car-info">
+          <div className="with-car-icon notification-icon">
+            <ServiceAlertIcon
+              className="inline-icon"
+              color="#007AC9"
+              severityLevel={AlertSeverityLevelType.Info}
+            />
+          </div>
+          <div className="with-car-info-notification">
```
Comment (vesameskanen) [inline app/component/itinerary/ItineraryListHeader.js:42]: merge these 2 divs: with-car-info-notification and with-car-info-content

File: app/component/itinerary/ItineraryPage.js [hunk 17]
```diff
@@ -1171,13 +1200,21 @@ export default function ItineraryPage(props, context) {
...
       />
     );
   }
 
+  const showCarPublicPlan =
+    settings.includeCarSuggestions &&
+    config.carBoardingModes?.FERRY !== undefined &&
+    carPublicPlan.carPublicItineraryCount > 0;
+
   const showAltBar =
```
Comment (vesameskanen) [inline app/component/itinerary/ItineraryPage.js:1309]: carPublicPlan.carPublicItineraryCount > 0 is enough

File: app/component/itinerary/ItineraryDetails.js [hunk 18]
```diff
@@ -133,21 +135,30 @@ class ItineraryDetails extends React.Component {
...
     const legsWithRentalBike = compressLegs(itinerary.legs).filter(leg =>
```
Comment (vesameskanen) [inline app/component/itinerary/ItineraryDetails.js:153]: walking.distance === 0 && biking.distance === 0 && driving.distance > 0

File: app/component/itinerary/ItineraryPage.js [hunk 19]
```diff
@@ -1171,13 +1200,21 @@ export default function ItineraryPage(props, context) {
         walking={walkPlan?.edges?.length > 0}
         biking={bikePlan?.edges?.length > 0 || !!bikePublicPlan?.edges?.length}
         driving={
-          (settings.includeCarSuggestions && carPlan?.edges?.length > 0) ||
+          (settings.includeCarSuggestions &&
+            (carPlan?.edges?.length > 0 ||
+              (config.carBoardingModes?.FERRY !== undefined &&
+                carPublicPlan?.edges?.length))) ||
```
Comment (vesameskanen) [inline app/component/itinerary/ItineraryPage.js]: Do not test FERRY. It is not necessary and also against idea of configurable boarding modes.

File: app/component/itinerary/ItineraryPage.js [hunk 20]
```diff
@@ -1187,7 +1224,10 @@ export default function ItineraryPage(props, context) {
       bikePlan?.edges?.length ||
       bikePublicPlan?.edges?.length ||
       parkRidePlan?.edges?.length ||
-      (settings.includeCarSuggestions && carPlan?.edges?.length));
+      (settings.includeCarSuggestions &&
+        (carPlan?.edges?.length ||
+          (config.carBoardingModes?.FERRY !== undefined &&
+            carPublicPlan?.edges?.length))));
```
Comment (vesameskanen) [inline app/component/itinerary/ItineraryPage.js]: (settings.includeCarSuggestions && carPlan?.edges?.length) ||
 carPublicPlan?.edges?.length > 0

**Review by vesameskanen — CHANGES_REQUESTED** [review]
Good work.  I suggested a fair amount of mainly cosmetic changes, which make the code easier to read and understand.

In general, try to avoid adding unncecessary validation checks 'just in case'.  The code should fail fast and crash if variable values are not valid.

File: app/component/RouteNumberContainer.js (hunk 1 above)
Comment (VillePihlava) [reply app/component/RouteNumberContainer.js]: Changed to this

File: app/component/RouteNumberContainer.js (hunk 2 above)
Comment (VillePihlava) [reply app/component/RouteNumberContainer.js]: I removed withCar (it now flows through props). Did the same to withBicycle

File: app/component/itinerary/CarLeg.js (hunk 3 above)
Comment (VillePihlava) [reply app/component/itinerary/CarLeg.js]: Changed

File: app/component/itinerary/CarLeg.js (hunk 4 above)
Comment (VillePihlava) [reply app/component/itinerary/CarLeg.js]: I did not change these for now. For example, there are similar divs in BicycleLeg, BikeParkLeg, CarParkLeg. Do you want all of these changed or none?

File: app/component/itinerary/CarLeg.js (hunk 5 above)
Comment (VillePihlava) [reply app/component/itinerary/CarLeg.js]: I did not change these for now. For example, there are similar divs in BicycleLeg, BikeParkLeg, CarParkLeg. Do you want all of these changed or none?

File: app/component/itinerary/InterlineInfo.js (hunk 6 above)
Comment (VillePihlava) [reply app/component/itinerary/InterlineInfo.js]: Done

File: app/component/itinerary/Itinerary.js (hunk 7 above)
Comment (VillePihlava) [reply app/component/itinerary/Itinerary.js]: I can find checks for leg.to in many parts of the code (e.g. CarParkLeg, CarLeg, TransitLeg, ViaLeg, WalkLeg). I assume therefore there is a chance for it to be null. Also the same code for bicycles used it in a similar way. Do you want me to look into it further or is this fine as is?

File: app/component/itinerary/ItineraryDetails.js (hunk 8 above)
Comment (VillePihlava) [reply app/component/itinerary/ItineraryDetails.js]: Fixed

File: app/component/itinerary/ItineraryListHeader.js (hunk 9 above)
Comment (VillePihlava) [reply app/component/itinerary/ItineraryListHeader.js]: This is the way it was for bikes. I set the default as an empty string for both

File: app/util/legUtils.js (hunk 10 above)
Comment (VillePihlava) [reply app/util/legUtils.js]: Changed

File: app/util/legUtils.js (hunk 11 above)
Comment (VillePihlava) [reply app/util/legUtils.js]: Changed

File: app/util/legUtils.js (hunk 12 above)
Comment (VillePihlava) [reply app/util/legUtils.js]: Changed

File: app/component/itinerary/ItineraryPageUtils.js (hunk 13 above)
Comment (VillePihlava) [reply app/component/itinerary/ItineraryPageUtils.js]: Changed

File: app/component/itinerary/ItineraryPageUtils.js (hunk 14 above)
Comment (VillePihlava) [reply app/component/itinerary/ItineraryPageUtils.js]: Changed

File: app/component/itinerary/ItineraryList.js (hunk 15 above)
Comment (VillePihlava) [reply app/component/itinerary/ItineraryList.js]: Removed

File: app/component/itinerary/ItineraryListHeader.js (hunk 16 above)
Comment (VillePihlava) [reply app/component/itinerary/ItineraryListHeader.js:42]: I removed with-bike-info-header and with-car-info-header in itinerary-list-header.scss, they were not used anymore. I merged with-bike-info-content and with-bike-info-notification as well as with-car-info-content and with-car-info-notification.

File: app/component/itinerary/ItineraryPage.js (hunk 17 above)
Comment (VillePihlava) [reply app/component/itinerary/ItineraryPage.js:1309]: FERRY checking is now removed. The idea behind checking for FERRY is that only in the configs that it is defined in, the UI will show ferries. Now there isn't a config field for removing carpublic itineraries in case they are found but not wanted. I did not remove settings.includeCarSuggestions.

File: app/component/itinerary/ItineraryPage.js (hunk 19 above)
Comment (VillePihlava) [reply app/component/itinerary/ItineraryPage.js]: FERRY checking is now removed. The idea behind checking for FERRY is that only in the configs that it is defined in, the UI will show ferries. Now there isn't a config field for removing carpublic itineraries in case they are found but not wanted.

File: app/component/itinerary/ItineraryPage.js (hunk 20 above)
Comment (VillePihlava) [reply app/component/itinerary/ItineraryPage.js]: This assumes that the carpublicplan would be shown even without the setting. I think this is wrong. Do you agree? FERRY is now removed

File: app/component/itinerary/CarLeg.js (hunk 4 above)
Comment (vesameskanen) [reply app/component/itinerary/CarLeg.js]: Ok I understand.  I would love to see all code improving continuously. Please fix this unless it feels unreasonably big task.

File: app/component/itinerary/CarLeg.js (hunk 5 above)
Comment (vesameskanen) [reply app/component/itinerary/CarLeg.js]: Same as above.

File: app/component/itinerary/Itinerary.js (hunk 7 above)
Comment (vesameskanen) [reply app/component/itinerary/Itinerary.js]: build/schema.graphql says in type Leg: 

 "The Place where the leg ends."
  to: Place!
! means that to cannot be null. Please fix at least your own code.

File: app/component/itinerary/ItineraryListHeader.js (hunk 16 above)
Comment (vesameskanen) [reply app/component/itinerary/ItineraryListHeader.js:42]: I noticed a new css class  .with-car-info-notification-icon which seems to be unused.

File: app/component/itinerary/ItineraryPage.js (hunk 17 above)
Comment (vesameskanen) [reply app/component/itinerary/ItineraryPage.js:1309]: Can there be car + public itineraries if includeCarSuggestions === false ? That would be surprising.

The logic gets a bit complicated because car itinerary is fetched for co2 comparison although user does not wish to get car itineraries.

Make sure that car + public won't get populated just because of co2 car routing. It should contain something only when user has selected car routing.

File: app/component/itinerary/ItineraryPage.js (hunk 19 above)
Comment (vesameskanen) [reply app/component/itinerary/ItineraryPage.js]: OK I will test this.  A more correct way would be testing if boardingModes is not empty, but I guess that is not needed. Code looks just  fine to me now.

**Comment by vesameskanen** [comment]
There is a conflict.

File: app/component/itinerary/ItineraryPage.js (hunk 20 above)
Comment (vesameskanen) [reply app/component/itinerary/ItineraryPage.js]: Combination (settings.includeCarSuggestions && carPlan?.edges?.length) is used because carplan may be comouted for co2 comparison only.

File: app/util/planParamUtil.js [hunk 21]
```diff
@@ -196,6 +203,13 @@ export function planQueryNeeded(
         settings.includeBikeSuggestions
       );
 
+    case PLANTYPE.CARTRANSIT:
+      return (
+        transitModes.length > 0 &&
+        settings.includeCarSuggestions &&
+        config.carBoardingModes?.FERRY !== undefined
```
Comment (vesameskanen) [inline app/util/planParamUtil.js]: This seems wrong. I suggest:  

 settings.includeCarSuggestions && filterTransitModes(...).length > 0

File: app/component/itinerary/ItineraryPage.js [hunk 22]
```diff
@@ -838,6 +850,20 @@ export default function ItineraryPage(props, context) {
     altStates[PLANTYPE.BIKETRANSIT][0].plan,
   ]);
 
+  // merge direct car and car transit plans into one
+  useEffect(() => {
+    if (
+      altStates[PLANTYPE.CAR][0].loading === LOADSTATE.DONE &&
+      altStates[PLANTYPE.CARTRANSIT][0].loading === LOADSTATE.DONE
+    ) {
+      const plan = mergeCarDirectAndTransitPlans(
```
Comment (vesameskanen) [inline app/component/itinerary/ItineraryPage.js:904]: I guess this is not OK. car + transit should be set only when there are carBoardingModes configured and settings.includeCarSuggestions is true.

**Review by vesameskanen — CHANGES_REQUESTED** [review]
I found some new logic errors related to car+public itinerary list creation. Please read my review comments.

File: app/component/itinerary/CarLeg.js (hunk 4 above)
Comment (VillePihlava) [reply app/component/itinerary/CarLeg.js]: These two divs have been merged everywhere I could find them.

File: app/component/itinerary/CarLeg.js (hunk 5 above)
Comment (VillePihlava) [reply app/component/itinerary/CarLeg.js]: These two divs have been merged everywhere I could find them.

File: app/component/itinerary/Itinerary.js (hunk 7 above)
Comment (VillePihlava) [reply app/component/itinerary/Itinerary.js]: I removed the ? from all uses of leg.to?.field and leg.from?.field I could find. However, in legUtils removing the check caused a test to fail so I added the ? back there

File: app/component/itinerary/ItineraryListHeader.js (hunk 16 above)
Comment (VillePihlava) [reply app/component/itinerary/ItineraryListHeader.js:42]: I removed with-car-info-notification-icon and with-bike-info-notification-icon. The car class was accidentally created based off of the old bike icon class. Both seem unused and I removed them

File: app/component/itinerary/ItineraryPage.js (hunk 17 above)
Comment (VillePihlava) [reply app/component/itinerary/ItineraryPage.js:1309]: I had a double check here and forgot I checked this earlier. This is now only carPublicPlan.carPublicItineraryCount > 0

File: app/component/itinerary/ItineraryPage.js (hunk 20 above)
Comment (VillePihlava) [reply app/component/itinerary/ItineraryPage.js]: This has now been changed to (settings.includeCarSuggestions && carPlan?.edges?.length) ||
carPublicPlan?.edges?.length > 0. I had a double check here and forgot I checked this earlier

File: app/util/planParamUtil.js (hunk 21 above)
Comment (VillePihlava) [reply app/util/planParamUtil.js]: I changed this to transitModes.length > 0 && settings.includeCarSuggestions. I didn't realize previously that transitModes is filtered earlier with filterTransitModes.

File: app/component/itinerary/ItineraryPage.js (hunk 22 above)
Comment (VillePihlava) [reply app/component/itinerary/ItineraryPage.js:904]: This is very similar to how the mergeBikeTransitPlans works. Because the check for making the cartransit query is made in planParamUtil, this should be fine as is. Do you want me to change this and the way it works for bikes or should I leave both as is?

Comment (VillePihlava) [reply app/component/itinerary/ItineraryPage.js:904]: I now changed it to
altStates[PLANTYPE.CAR][0].loading === LOADSTATE.DONE &&
altStates[PLANTYPE.CARTRANSIT][0].loading === LOADSTATE.DONE &&
settings.includeCarSuggestions &&
config.carBoardingModes !== undefined

File: app/component/itinerary/Itinerary.js (hunk 7 above)
Comment (vesameskanen) [reply app/component/itinerary/Itinerary.js]: right, unit tests often use bad mock data. Fixing code causes huge amount of test failures which is really annoying.

**Review by vesameskanen — CHANGES_REQUESTED** [review]
Looks good now but merge confict must be resolved.

**Comment by VillePihlava** [comment]
Merge conflicts resolved

**Review by vesameskanen — APPROVED** [review]
Will be merged after next hotfix release
